File: backend/src/cortex/domain/document.py
from __future__ import annotations

import enum
from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path
from uuid import UUID, uuid4


class FileType(str, enum.Enum):
    PDF = "pdf"
    MARKDOWN = "markdown"
    DOCX = "docx"
    XLSX = "xlsx"
    TXT = "txt"
    PNG = "png"
    JPG = "jpg"
    TIFF = "tiff"


class ProcessingStatus(str, enum.Enum):
    UPLOADING = "uploading"
    STORED = "stored"
    PARSING = "parsing"
    PARSED = "parsed"
    CHUNKING = "chunking"
    CHUNKED = "chunked"
    EMBEDDING = "embedding"
    EMBEDDED = "embedded"
    EXTRACTING_ENTITIES = "extracting_entities"
    ENTITIES_EXTRACTED = "entities_extracted"
    BUILDING_GRAPH = "building_graph"
    READY = "ready"
    FAILED = "failed"


@dataclass
class DocumentMetadata:
    page_count: int | None = None
    word_count: int | None = None
    language: str = "en"
    author: str | None = None
    subject: str | None = None


@dataclass
class ParseResult:
    text: str
    structured: dict
    rendered_html: str
    rendered_markdown: str
    metadata: DocumentMetadata
    images: list[ExtractedImage] = field(default_factory=list)
    thumbnail_path: str | None = None
    page_count: int | None = None


@dataclass
class ExtractedImage:
    image_path: str
    page_number: int | None = None
    caption: str | None = None
    alt_text: str | None = None
    width: int | None = None
    height: int | None = None


@dataclass
class Document:
    id: UUID
    title: str
    original_filename: str
    file_type: FileType
    file_size_bytes: int
    file_hash: str
    mime_type: str
    original_path: str
    status: ProcessingStatus = ProcessingStatus.UPLOADING
    thumbnail_path: str | None = None
    parsed_content: dict | None = None
    rendered_markdown: str | None = None
    rendered_html: str | None = None
    metadata: DocumentMetadata = field(default_factory=DocumentMetadata)
    error_message: str | None = None
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    processed_at: datetime | None = None
    collection_id: UUID | None = None
    tags: list[str] = field(default_factory=list)
    is_favorite: bool = False

    @staticmethod
    def new(
        title: str,
        original_filename: str,
        file_type: FileType,
        file_size_bytes: int,
        file_hash: str,
        mime_type: str,
        original_path: str,
    ) -> Document:
        return Document(
            id=uuid4(),
            title=title,
            original_filename=original_filename,
            file_type=file_type,
            file_size_bytes=file_size_bytes,
            file_hash=file_hash,
            mime_type=mime_type,
            original_path=original_path,
        )

File: frontend/Sources/Domain/Document.swift
import Foundation

package enum FileType: String, Sendable, Codable, CaseIterable {
    case pdf
    case markdown
    case docx
    case xlsx
    case txt
    case png
    case jpg
    case tiff

    package var iconName: String {
        switch self {
        case .pdf: "doc.richtext"
        case .markdown: "doc.text"
        case .docx: "doc.fill"
        case .xlsx: "tablecells"
        case .txt: "doc.plaintext"
        case .png, .jpg, .tiff: "photo"
        }
    }

    package var displayName: String {
        switch self {
        case .pdf: "PDF"
        case .markdown: "Markdown"
        case .docx: "Word"
        case .xlsx: "Excel"
        case .txt: "Text"
        case .png: "PNG"
        case .jpg: "JPEG"
        case .tiff: "TIFF"
        }
    }
}

package enum ProcessingStatus: String, Sendable, Codable {
    case uploading
    case stored
    case parsing
    case parsed
    case chunking
    case chunked
    case embedding
    case embedded
    case extractingEntities = "extracting_entities"
    case entitiesExtracted = "entities_extracted"
    case buildingGraph = "building_graph"
    case ready
    case failed

    package var isProcessing: Bool {
        switch self {
        case .ready, .failed:
            false
        default:
            true
        }
    }

    package var displayLabel: String {
        switch self {
        case .uploading: "Uploading..."
        case .stored: "Queued"
        case .parsing: "Parsing..."
        case .parsed: "Parsed"
        case .chunking: "Chunking..."
        case .chunked: "Chunked"
        case .embedding: "Embedding..."
        case .embedded: "Embedded"
        case .extractingEntities: "Extracting entities..."
        case .entitiesExtracted: "Entities extracted"
        case .buildingGraph: "Building graph..."
        case .ready: "Ready"
        case .failed: "Failed"
        }
    }
}

package struct Document: Sendable, Equatable, Codable, Identifiable {
    package let id: UUID
    package var title: String
    package let originalFilename: String
    package let fileType: FileType
    package let fileSizeBytes: Int
    package let mimeType: String
    package var status: ProcessingStatus
    package var pageCount: Int?
    package var wordCount: Int?
    package var language: String?
    package var author: String?
    package var tags: [String]
    package var isFavorite: Bool
    package var collectionId: UUID?
    package let createdAt: Date
    package var updatedAt: Date
    package var processedAt: Date?
    package var errorMessage: String?

    enum CodingKeys: String, CodingKey {
        case id, title, status, tags, language, author
        case originalFilename = "original_filename"
        case fileType = "file_type"
        case fileSizeBytes = "file_size_bytes"
        case mimeType = "mime_type"
        case pageCount = "page_count"
        case wordCount = "word_count"
        case isFavorite = "is_favorite"
        case collectionId = "collection_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case processedAt = "processed_at"
        case errorMessage = "error_message"
    }
}

package struct DocumentListResponse: Sendable, Codable {
    package let documents: [Document]
    package let total: Int
    package let limit: Int
    package let offset: Int
}

package struct DocumentUploadResponse: Sendable, Codable {
    package let id: UUID
    package let status: String
    package let message: String
    package let isDuplicate: Bool

    enum CodingKeys: String, CodingKey {
        case id, status, message
        case isDuplicate = "is_duplicate"
    }
}

package struct DocumentContent: Sendable, Codable {
    package let id: UUID
    package let format: String
    package let content: String
    package let originalUrl: String
    package let metadata: Document

    enum CodingKeys: String, CodingKey {
        case id, format, content, metadata
        case originalUrl = "original_url"
    }
}
